### ISSUE TITLE
Tensorboard with pytorch

### ISSUE THREAD

**Issue by bbacem80** [PacktPublishing/Hands-On-Intelligent-Agents-with-OpenAI-Gym#14]
Hello @praveen-palanisamy 

I went through your book and it's really helping.
I have a question regarding "Using TensorBoard for logging and visualizing a PyTorch RL agent's progress" p. 107 (chap 6). Which tool did you use to generate the graph reflecting the algorithm performance? While Tensorboard is more or less evident to use in Tensorflow, I haven't been able to find an equivalent tool in Pytorch.

Thank you in advance.

**Comment by praveen-palanisamy** [comment]
Hi @bbacem80 , 
Glad that you find the book to be helpful.

I indeed used Tensorboard to generate the graph showing the agent's performance. It is true that Tensorboard is closely integrated with TensorFlow but since Tensorboard can be used as a standalone tool and has an easy to use I/O,  there are utility/libraries that can log data in the format that Tensorboard can parse and visualize.

For logging PyTorch data that can be visualized using Tensorboard, there are a few tools but the one I used (which is also what is indicated as part of the [book's python environment setup here](https://github.com/PacktPublishing/Hands-On-Intelligent-Agents-with-OpenAI-Gym/blob/master/conda_env.yaml)) is [TensorboardX](https://github.com/lanpa/tensorboardX) which you can install using: `pip install tensorboardX`.

For an usage example, you can refer to [this script](), part of chapter 6. Following is a simple snippet showing the usage:
https://github.com/PacktPublishing/Hands-On-Intelligent-Agents-with-OpenAI-Gym/blob/34adb43a9d419e0de5472b268fe6102debccded6/ch6/deep_Q_learner.py#L109-L117

Line 111 in the snippet above logs the epsilon exploration decay value as the training progresses which can be seen on Tensorboard. `writer` is a `TensorboardX.SummaryWriter` object.

Hope that helps.

**Comment by bbacem80** [comment]
Hello @praveen-palanisamy,

I'm now able to add scalars and generate log files under the directory /logs/A2C_Carla-v0_19-01-22-11-33 (e.g.). Once TensorBoard installed, I need just to open one of the files using

tensorboard --logdir=/logs/A2C_Carla-vX

?

Regards,

**Comment by YasminaJaafra** [comment]
Here I'm in the case of the policy (CNN weights and biases) re-initialization (e.g. at the beginning of each trajectory execution).
I assume there are 3 ways to operate it:
1- Randomly
2- Using files to save - load a specific model
3- As a variable in the code.
Do you have an idea about how implementing the 3rd option, i.e. save a model as a variable and call it to initialize a CNN when needed.

Regards,

**Comment by YasminaJaafra** [comment]
Hello @praveen-palanisamy,

I think that this point could be related to the policy (CNN weights and biases) re-initialization (e.g. at the beginning of each trajectory execution). So I have the request below:

I assume that there are 3 ways to operate CNN model (weights and biases) re-initialization:
1- Randomly
2- Using files to save - load a specific model
3- As a variable in the code.

Do you have an idea about how implementing the 3rd option, i.e. save a model as a variable and call it to initialize a CNN when needed?

Best Regards,

**Comment by praveen-palanisamy** [comment]
@YasminaJaafra:
For the 3rd option, let's say you have the CNN model (or any `nn.model`) stored in a variable/object named `model1`, you can access the parameters (weights and biases) using `model1.parameters()`if you want to initialize another instance of the  model with the same parameters (weights and biases) as of `model1`,  you can do either of the following:
  a. Iterate over `model1.parameters()` and copy all the parameters layer by layer or
  b. Copy the `state_dict` (which is a dict containing all the learn-able parameters of `model1`) of the model

Option b. is very similar to how the 2- nd option (using files to save - load) is implemented in the code. For example, to initialize `model2` with the same weights and biases as `model1`, you can use the following line of code:
```python
model2.load_state_dict(model1.state_dict())
```

You can find modular example in the Asynchronous Advantage Actor Critic agent implementation's global sync code in Chapter 8, where the parameters are synced using the `state_dict` as shown below where the weights of `self.actor` and `self.critic` are loaded from another (global) variable/object.

https://github.com/PacktPublishing/Hands-On-Intelligent-Agents-with-OpenAI-Gym/blob/df9ab3984237b3a02998e2c3d3df482f557945f9/ch8/async_a2c_agent.py#L198-L213

Hope that helps.
If you have a follow-up question, please move this to a new issue as #14 was opened for discussing about a completely different topic.
